plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,564 @@
+                .bucket(bucketName)
+                .key(key)
+                .build();
+        s3Client.deleteObject(request);
+    }
+
+    private static String getBucketName(URI uri)
+    {
+        if (uri.getHost() != null) {
+            return uri.getHost();
+        }
+
+        if (uri.getUserInfo() == null) {
+            return uri.getAuthority();
+        }
+
+        throw new IllegalArgumentException("Unable to determine S3 bucket from URI.");
+    }
+
+    private static String keyFromUri(URI uri)
+    {
+        checkArgument(uri.isAbsolute(), "Uri is not absolute: %s", uri);
+        String key = nullToEmpty(uri.getPath());
+        if (key.startsWith(PATH_SEPARATOR)) {
+            key = key.substring(PATH_SEPARATOR.length());
+        }
+        if (key.endsWith(PATH_SEPARATOR)) {
+            key = key.substring(0, key.length() - PATH_SEPARATOR.length());
+        }
+        return key;
+    }
+
+    private static boolean isDirectory(URI uri)
+    {
+        return uri.toString().endsWith(PATH_SEPARATOR);
+    }
+
+    private static AwsCredentialsProvider createAwsCredentialsProvider(ExchangeS3Config config)
+    {
+        if (config.getS3AwsAccessKey() != null && config.getS3AwsSecretKey() != null) {
+            return StaticCredentialsProvider.create(AwsBasicCredentials.create(config.getS3AwsAccessKey(), config.getS3AwsSecretKey()));
+        }
+        return DefaultCredentialsProvider.create();
+    }
+
+    private S3Client createS3Client(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3ClientBuilder clientBuilder = S3Client.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3AsyncClientBuilder clientBuilder = S3AsyncClient.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private static class S3ExchangeStorageWriter
+            implements ExchangeStorageWriter
+    {
+        private final S3Client s3Client;
+        private final S3AsyncClient s3AsyncClient;
+        private final String bucketName;
+        private final String key;
+        private final int partSize;
+        private final Optional<SecretKey> secretKey;
+
+        private boolean failed;
+        private Optional<String> uploadId = Optional.empty();
+        private CompletableFuture<UploadPartResponse> inProgressUploadFuture;
```

Comment:
`uploadInProgressFuture`?

Reply:
This naming imitates `TrinoS3FileSystem`